staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1217,13 +1227,41 @@ var _ runtime.Decoder = schemaCoercingDecoder{}
 
 func (d schemaCoercingDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
 	obj, gvk, err := d.delegate.Decode(data, defaults, into)
+	var decodingStrictErrs []error
 	if err != nil {
-		return nil, gvk, err
+		if !runtime.IsStrictDecodingError(err) {
+			return nil, gvk, err
+		}
+		decodeStrictJSONErr, ok := runtime.AsStrictDecodingError(err)
+		if !ok {
+			// unreachable
+			return nil, gvk, fmt.Errorf("error both is and is not a strict decoding error: %v", err)
+		}
+		decodingStrictErrs = decodeStrictJSONErr.Errors()
 	}
 	if u, ok := obj.(*unstructured.Unstructured); ok {
-		if err := d.validator.apply(u); err != nil {
+		unknownFields, preserved, err := d.validator.apply(u)
+		if err != nil {
 			return nil, gvk, err
 		}
+		if len(unknownFields) > 0 && d.validator.returnUnknownFieldPaths {
+			applyStrictErrs := make([]error, len(unknownFields))
+			for i, unknownField := range unknownFields {
+				applyStrictErrs[i] = fmt.Errorf(`unknown field "%s"`, unknownField)
+			}
+			// when apply indicates that preserved is true,
+			// we need to return a special type of strict decoding error
+			// (PreservedDecodingError) so that the handler knows
+			// that it should warn instead of error in the strict case
+			// in order to maintain compatibility with how
+			// preserve unknown fields was handled prior to server-side
+			// validation.
+			if preserved {
+				return obj, gvk, runtime.NewPreservedDecodingError(append(decodingStrictErrs, applyStrictErrs...))
+			}
+			return obj, gvk, runtime.NewStrictDecodingError(append(decodingStrictErrs, applyStrictErrs...))
```

Comment:
I believe we are doing exactly what you are describing, appending to `decodingStrictErrs` here (see the value getting passed to `NewStrictDecodingError` is the output of the append). 

I’m not sure what you mean by returning early, because we aren’t inside of a loop here, this is just collecting the errors when it’s appropriate to (e.g. that it is an unstructured and apply does return strict errors) 

You can see that this is tested in the `TestFieldValidationPostCRD/crd-post-strict-validation`.It should be easier to read once I get rid of the preserved decoding errors junk, but let me know if I’m missing anything about your comment.